packages/loader/container-loader/src/connectionHandler.ts

```diff
@@ -0,0 +1,889 @@
+        details.checkpointSequenceNumber = checkpointSequenceNumber;
+        this.props.connectHandler(details);
+
+        this.connectFirstConnection = false;
+    }
+
+    /**
+     * Disconnect the current connection and reconnect.
+     * @param connection - The connection that wants to reconnect - no-op if it's different from this.connection
+     * @param requestedMode - Read or write
+     * @param error - Error reconnect information including whether or not to reconnect
+     * @returns A promise that resolves when the connection is reestablished or we stop trying
+     */
+     private reconnectOnError(
+        requestedMode: ConnectionMode,
+        error: DriverError,
+    ) {
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        this.reconnectOnErrorCore(
+            requestedMode,
+            error.message,
+            error);
+    }
+
+    /**
+     * Disconnect the current connection and reconnect.
+     * @param connection - The connection that wants to reconnect - no-op if it's different from this.connection
+     * @param requestedMode - Read or write
+     * @param error - Error reconnect information including whether or not to reconnect
+     * @returns A promise that resolves when the connection is reestablished or we stop trying
+     */
+    private async reconnectOnErrorCore(
+        requestedMode: ConnectionMode,
+        disconnectMessage: string,
+        error?: DriverError,
+    ) {
+        // We quite often get protocol errors before / after observing nack/disconnect
+        // we do not want to run through same sequence twice.
+        // If we're already disconnected/disconnecting it's not appropriate to call this again.
+        assert(this.connection !== undefined, 0x0eb /* "Missing connection for reconnect" */);
+
+        this.disconnectFromDeltaStream(disconnectMessage);
+
+        const canRetry = error !== undefined ? canRetryOnError(error) : true;
+
+        // If reconnection is not an option, close the DeltaManager
+        if (this.reconnectMode === ReconnectMode.Never || !canRetry) {
+            // Do not raise container error if we are closing just because we lost connection.
+            // Those errors (like IdleDisconnect) would show up in telemetry dashboards and
+            // are very misleading, as first initial reaction - some logic is broken.
+            this.props.closeHandler(canRetry ? undefined : error);
+        }
+
+        // If closed then we can't reconnect
+        if (this.closed || this.reconnectMode !== ReconnectMode.Enabled) {
+            return;
+        }
+
+        const delayMs = error !== undefined ? getRetryDelayFromError(error) : undefined;
+        if (delayMs !== undefined) {
+            this.props.reconnectionDelayHandler(delayMs, error);
+            await waitForConnectedState(delayMs);
+        }
+
+        this.triggerConnect(requestedMode);
+    }
+
+    /**
```

Comment:
This comment does not belong to this method. Looks like it was copied from submit but doesn't apply here.